File: pset6/crack/crack.py
from sys import argv
import crypt
from cs50 import get_string


def CharIncrement(char, depth):

    # how to change a letter?

    if (char[depth] >= 'A' and char[depth] < 'Z') or (char[depth] >= 'a' and char[depth] < 'z'):
        char = char[:depth] + chr(ord(char[depth]) + 1) + char[depth + 1:]
    elif char[depth] == 'Z':
        char = char[:depth] + 'a' + char[depth + 1:]
    else:
        char = char[:depth] + 'A' + char[depth + 1:]
        if depth + 1 == len(char):
            char += 'A'
        char = CharIncrement(char, depth + 1)
    return char


if not len(argv) == 2:
    print("Usage: ./crack key")
    exit(1)

else:
    pass_hash = argv[1]
    salt = pass_hash[0] + pass_hash[1]

    pass_code = "A"
    while True:
        if crypt.crypt(pass_code, salt) == pass_hash:
            break
        else:
            pass_code = CharIncrement(pass_code, 0)

    print(pass_code)

File: pset2/crack.c
#include <cs50.h>
#include <stdio.h>
#include <crypt.h>
#include <string.h>

void CharIncrement(string pass, int depth);

int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Usage: ./crack hash");
        return 1;
    }
    string hash = argv[1];
    char salt[] = "AA";
    salt[0] = hash[0];
    salt[1] = hash[1];
//    printf("Salt: %s\n", salt);
    char pass[] = {'A', '\0', '\0', '\0', '\0'};
    while (crypt(pass, salt) != hash/* && pass[1] == '\0'*/)
    {
        CharIncrement(pass, 0);
//        printf("pass: %s, hash: %s\n", pass, crypt(pass, salt));
    }
    printf("%s\n", pass);
    return 0;
}

void CharIncrement(char pass[], int depth)
{
    if ((pass[depth] >= 'A' && pass[depth] < 'Z') || (pass[depth] >= 'a' && pass[depth] < 'z'))
    {
        // increment the character
        pass[depth]++;
        return;
    }
    else if (pass[depth] == 'Z')
    {
        pass[depth] = 'a';
        return;
    }
    else
    {
        pass[depth] = 'A';
        if (pass[depth + 1] == '\0')
        {
            pass[depth + 1] = 'A';
            return;
        }
        else
        {
            CharIncrement(pass, depth + 1);
            return;
        }
    }
}
